File: leetcode/editor/cn/234_回文链表_test.go
package leetcode

import (
	"github.com/stretchr/testify/assert"
	"testing"
)

func Test_givenNormal_when234_thenSuccess(t *testing.T) {
	assert.Equal(t, true, isPalindrome(NewList([]int{1, 2, 3, 2, 1})))
	assert.Equal(t, true, isPalindrome(NewList([]int{1, 2, 2, 1})))
	assert.Equal(t, false, isPalindrome(NewList([]int{1, 2})))
}

//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func isPalindrome(head *ListNode) bool {
	// length
	length := 0
	for ptr := head; ptr != nil; ptr = ptr.Next {
		length += 1
	}
	if length < 2 {
		return true
	}
	// 不用考虑长度为奇数或偶数
	ptr := head
	for i := 0; i < (length-1)/2; i++ {
		ptr = ptr.Next
	}
	// 反转后半段链表
	var pre *ListNode = nil
	for cur := ptr.Next; cur != nil; {
		next := cur.Next
		cur.Next = pre
		pre = cur
		cur = next
	}
	// 双指针
	ans := true
	left, right := head, pre
	for right != nil {
		if left.Val != right.Val {
			ans = false
			break
		}
		left, right = left.Next, right.Next
	}
	return ans
}

//leetcode submit region end(Prohibit modification and deletion)

File: leetcode/editor/cn/234_回文链表_test.py
import unittest
from typing import Optional

from model.list_node import ListNode


class TestSolution(unittest.TestCase):

    def setUp(self) -> None:
        self.solution = Solution()

    def test_givenNormal_when_thenSuccess(self):
        palindrome = ListNode.new_list([1, 2, 3, 2, 1])
        self.assertTrue(self.solution.isPalindrome(palindrome))
        #
        palindrome = ListNode.new_list([1, 2, 2, 1])
        self.assertTrue(self.solution.isPalindrome(palindrome))
        #
        not_palindrome = ListNode.new_list([1, 2])
        self.assertFalse(self.solution.isPalindrome(not_palindrome))

# leetcode submit region begin(Prohibit modification and deletion)
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if head is None or head.next is None:
            return True
        stack = []
        slow, fast = head, head.next
        while fast is not None and fast.next is not None:
            slow = slow.next
            fast = fast.next.next
        # slow is the first half
        ptr = head
        while ptr != slow.next:
            stack.append(ptr.val)
            ptr = ptr.next
        # 奇数长度链表
        if fast is None:
            stack.pop()
        while ptr is not None and ptr.val == stack.pop():
            ptr = ptr.next
        return ptr is None


# leetcode submit region end(Prohibit modification and deletion)


if __name__ == '__main__':
    unittest.main()
